### PR TITLE
Enhancing Quote block styling for different alignments

### LINKED ISSUE #11609: Right-align border for right-aligned quotes
To improve the consistency of alignments and to avoid creating potentially-awkward rivers inside the quote block, I'd recommend adjusting the positioning of the hard left border on the quote block when the alignment is changed, so it would be to the right on a right-aligned quote, and either removed or maybe at the top of a centre-aligned block.

**Left-aligned**

screenshot 2018-11-08 01 07 40

Suggestion is to remain as-is, since the alignment here is consistent with the text.

**Centre-aligned**

screenshot 2018-11-08 01 07 29

Suggestion:
screenshot 2018-11-08 01 08 57

**Right-aligned**
screenshot 2018-11-08 01 07 50

Suggestion:
screenshot 2018-11-08 01 08 17

### PR DESCRIPTION
**Pull request by colorful-tones**
Fixes #11609 _and_ addresses #10970 

## Description
These changes affect the Quote block. 
* Added: className assignment based on alignment chosen in BlockControls (`.is-left` (default), `.is-center`, or `is-right`) for editor and frontend output.
* Enhanced: styling for `blockquote` based on `is-left`, `is-center`, or `is-right`m while also accounting for the existing Style option of either: "Regular" or "Large" quotes.

## How has this been tested?
Changes were tested in local environment with Gutenberg post with all variations of Quote block added. Changes were tested against the TwentySixteen theme _and_ TwentyNineteen theme (currently being developed).

Note: in TwentyNineteen there are already some sane and simple stylistic overrides for Gutenberg Quotes. This PR's changes has a negative stylistic impact on the current styles in TwentyNineteen. However, I would be happy to open and submit a PR to address this minor updates to align with this set of changes.

Tested in Chrome: 70.0.3538.77 on macOS 10.14

## Screenshots 

twentysixteen-gutenberg-quotes-block

twentynineteen-gutenberg-quote-block

Gutenberg editor new Quote block styles

## Types of changes
Previously the Quote block had the ability to use the BlockControls to align: left, center or right. However, the alignment was being set on the `<blockquote>` element with `style="text-align: right;"`. I removed this assignment and replaced with className assignment.

| Before                        | After             |
|-------------------------------|-------------------|
| default/none                  | class="is-left"   |
| `style="text-align: right;"`   | class="is-right"  |
| `style="text-align: center;"` | class="is-center" |

## Checklist:
- [ x ] My code is tested.
- [ x ] My code follows the WordPress code style. 
- [ x ] My code follows the accessibility standards. 
- [ x ] My code has proper inline documentation. 

(via @10up)

### CODE REVIEW

**Review by sarahmonster — CHANGES_REQUESTED** [review]
Hi @colorful-tones! Thanks for the pull request. I've gone through and left some comments, but overall it looks good. 

Whilst I think this is an overall improvement, it may not be something we can merge just yet due to our UI freeze. I'm going to loop in @jasmussen as well to take a look at some of the CSS issues, since he'll likely have a much better concept of any potential conflicts there.

File: packages/block-library/src/quote/editor.scss [hunk 1]
```diff
@@ -1,7 +1,47 @@
...
+		padding-left: 1em;
+	} // &.is-left
+
+	&.is-right {
+		border-right: 4px solid $black;
+		padding-right: 1em;
+		text-align: right;
+	} // &.is-right
+
+	&.is-center {
```
[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/editor.scss]: I disagree. I do not believe there is anything over complicated about the CSS using pseudo-elements. I believe it affords a nicer, elegant out-of-box look for centered quotes, while also allowing a lot of options for theme authors to overrides `border-color`, `border-width`, `width` and more.

File: packages/block-library/src/quote/editor.scss [hunk 2]
```diff
@@ -1,7 +1,47 @@
 .wp-block-quote {
-	margin: 0;
+	margin: 0 inherit;
```
[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/editor.scss]: I had to be presumptuous here. I've seen too many inexperienced CSS authors be overly explicit in learning to write CSS with shorthand properties. 

Further reading: [Harry Roberts "CSS Shorthand Syntax Considered an Anti-Pattern"](https://csswizardry.com/2016/12/css-shorthand-syntax-considered-an-anti-pattern/)

The left/right margin seemed overly explicit, unnecessary, and would cause more headaches for theme authors to overwrite with `!important` further down the line. I agree that there is potential for "unanticipated regressions" now. However, I project that the possibility of these regressions now within the Gutenberg editor vs the outputted styling and front-end experience would outweigh the potential to allow this overly explicit styling to remain.

File: packages/block-library/src/quote/index.js [hunk 3]
```diff
@@ -33,6 +34,7 @@ const blockAttributes = {
 	},
 	align: {
 		type: 'string',
+		default: 'is-left',
```
[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/index.js]: Hmm, the `default: 'is-left'` is saying that newly created Quote blocks should automatically have the `.is-left` class assigned, therefore left-aligned styling applied.

This was the case before, as newly created Quote blocks had `style="text-align: left;"` automatically applied. So, nothing changes really. Should an editor re-open a previously saved left-aligned quote block it would not suddenly lose its styling. Instead, `style="text-align: left;"` would be removed and `.is-left` would get substituted.

With that said, it could potentially have regressions on prior blocks that were saved as `style="text-align: right;"` or `style="text-align: center;"`, because it would replace _those_ with `.is-left` automatically. 

I could likely work up an instance using the [`deprecated` block API](https://wordpress.org/gutenberg/handbook/block-api/deprecated-blocks/) to allow for graceful transition.

File: packages/block-library/src/quote/index.js [hunk 4]
```diff
@@ -189,17 +191,23 @@ export const settings = {
...
+		);
+
 		return (
 			<Fragment>
 				<BlockControls>
 					<AlignmentToolbar
 						value={ align }
 						onChange={ ( nextAlign ) => {
-							setAttributes( { align: nextAlign } );
+							setAttributes( { align: 'is-' + nextAlign } );
```
[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/index.js]: Yeah, I'm not certain what the conventions are here either. I did peruse some other blocks, and based on things like `.is-large`, `.is-style-large` in current Quote block: https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/quote/theme.scss

and `.is-style-solid-color` in current Pullquote block: https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/pullquote/editor.scss#L30

I extrapolated: `.is-left`, `.is-center` and `.is-right`. I think these could potentially be reused (and likely should be) within other blocks as well. In this particular instance I don't get a sense that namespacing is typically used from what I've seen of Gutenberg codebase.

File: packages/block-library/src/quote/style.scss [hunk 5]
```diff
@@ -1,19 +1,19 @@
 .wp-block-quote {
+
 	&.is-style-large,
 	&.is-large {
-		margin: 0 0 16px;
```
[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/style.scss]: So, this is interesting to me, because currently in Gutenberg Quote block's `theme.scss` we're assigning a `margin-bottom` of `20px` on the overall `.wp-block-quote` wrapper. However, when the Quote block (again, current Gutenberg treatment) gets assigned as `.is-large` by the user, than we're _decreasing_ the `margin-bottom` from `20px` to `16px` here: https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/quote/style.scss#L2

This screams inconsistency in oversight to me. 😞 

Therefore, this is why I went to the extent of trying to update the overall styling of the Quote block component to a little more consistent and scaleable solution. In hopes that it would get merged before Gutenberg launch (WP 5.0).

Also, to clarify I changed `px` units to `em`, because using `px` is making things far too static, and has potential a11y concerns for users on both the back end and the front end.

File: packages/block-library/src/quote/theme.scss [hunk 6]
```diff
@@ -1,5 +1,5 @@
 .wp-block-quote {
-	margin: 20px 0;
+	margin: 1em inherit;
```
[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/theme.scss]: Please see previous reasoning:
1. here: https://github.com/WordPress/gutenberg/pull/11646/files#r232811114 and
2. here: https://github.com/WordPress/gutenberg/pull/11646/files#r232819630

File: packages/block-library/src/quote/editor.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/editor.scss]: @sarahmonster thanks for the great feedback. I truly appreciate it.

I tried to provide further reasoning behind my decisions. I would hope and encourage that we get another party's insight. I'm still new to the Gutenberg codebase and I don't want to miss anything in the larger scheme of things.

I hope that my reasoning is not perceived as a "no". I'm mostly just curious what more people think as I'm new to the project and tried to apply my most sound reasoning and experience, while also perusing existing Issues, Git Blame, and general history to get a sense of best practices and approaches. 

Hopefully, somebody else will drop some feedback and I can get all this updated in another pass, sooner than later. Thanks again!

[the remark this answers is not in the document]
Comment (jasmussen) [reply packages/block-library/src/quote/editor.scss]: I would agree with Sarah here that this is a little sensitive to introduce at this point, which is right before the 5.0 release. Which means we have to be extra careful not to cause regressions. In this case it's important to test this both with themes that _do_ support wide alignments, and with themes that do not. There's also mobile, tablet, desktop breakpoints to consider, and even though these are editor styles, the frontend should ideally match. 

I understand your arguments and reasons for changing this — and I think this would make a good separate pull request that looks at these editor specific margin rules for _all the blocks, together_. But changing this for just the quote block in isolation is just going to cause inconsistency, even if this becomes the only block that's "right". 

In other words, unless this rule is _strictly necessary_ for the purpose of this PR, I would not bundle it into this PR.

File: packages/block-library/src/quote/editor.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (jasmussen) [reply packages/block-library/src/quote/editor.scss]: I have a small preference for Sarahs design, and would generally agree that simpler is almost always better. Additionally, the double borders makes it look more like the Pullquote block:

screenshot 2018-11-13 at 11 37 11

Your argument is not wrong, and there's nothing _wrong_ with what you're doing here. In other words, both your points are valid.

Separate note on this, this is how Twenty Nineteen looks now:

screenshot 2018-11-13 at 11 30 35

screenshot 2018-11-13 at 11 30 41

In that vein it's worth clarifying a thing:

1. editor.scss is always loaded by the editor
2. style.scss is always loaded by the editor, and is always loaded by the frontend
3. theme.scss is only loaded in editor and frontend if the theme opts into Gutenberg "opinionated styles"

TwentyNineteen opts into those opinionated styles. And those opinionated styles exist exactly so we can do things like these. But this also means we need to move all those visual styles into theme.scss so a theme only gets them if opting in. Right now there are a lot of rules in editor.scss that will be present regardless of theme opt-in.

File: packages/block-library/src/quote/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jasmussen) [reply packages/block-library/src/quote/index.js]: This is not a change I'd recommend we do this close to 5.0 release, and it also needs to take RTL languages into account.

File: packages/block-library/src/quote/style.scss (hunk 5 above)
[the remark this answers is not in the document]
Comment (jasmussen) [reply packages/block-library/src/quote/style.scss]: I deeply appreciate your work here, and while you have several valid points, as we barrel towards the November 27th release, it's critical that we focus pull requests only on solving very specific issues, and that we avoid lumping in unrelated changes. 

Even though there can be an argument for em changes, it's unrelated to the initial issue, and is only making this review more complex.

> So, this is interesting to me, because currently in Gutenberg Quote block's theme.scss we're assigning a margin-bottom of 20px on the overall .wp-block-quote wrapper (see line #2). However, when the Quote block (again, current Gutenberg treatment) gets assigned as .is-large by the user, than we're decreasing the margin-bottom from 20px to 16px here: 

Solid observation, worth fixing.

File: packages/block-library/src/quote/editor.scss [hunk 7]
```diff
@@ -1,7 +1,47 @@
...
+
+		&::before {
+			background-color: $black;
+			content: "";
+			display: block;
+			height: 4px;
+			margin-left: calc(50% - 10%);
+			margin-bottom: 2em;
+			width: 20%;
+		} // &::before
```
Comment (jasmussen) [inline packages/block-library/src/quote/editor.scss]: Let's not do these closing comments. They aren't added anywhere else in Gutenberg.

**Review by jasmussen — CHANGES_REQUESTED** [review]
Thank you so much for this PR! I really dig the style change Sarah proposed, and thank you for taking a stab at the implementation.

I left a few minor comments around the design implementation. But overall the changes that need to happen for this PR to ship, notably if we mean to ship it soon, are all related to making the entire PR smaller and more focused. Although you have many good arguments, to me those are all separate PRs that look at Gutenberg overall instead of the quote block in isolation. As it is the changes added here are likely to require a deprecation handler, and require testing for RTL, wide/fullwide, mobile, and theme editor styles.

**Comment by jasmussen** [comment]
I expanded the review range to get you more help in shipping a version of this!

File: packages/block-library/src/quote/editor.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/editor.scss]: @jasmussen @sarahmonster

Thanks for the thorough and insightful review!

One thing still remains unclear. There are several mentions of “causing regressions”. Perhaps I’m missing some oversight on the Gutenberg project’s milestones and project goals. Since it has not been “officially” released yet then why are we concerned with regressions?

File: packages/block-library/src/quote/editor.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/editor.scss]: Yes, all these TwentyNineteen issues were noted in the original PR notes above. Thanks for clarifying though.

File: packages/block-library/src/quote/editor.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasmussen) [reply packages/block-library/src/quote/editor.scss]: > One thing still remains unclear. There are several mentions of “causing regressions”. Perhaps I’m missing some oversight on the Gutenberg project’s milestones and project goals. Since it has not been “officially” released yet then why are we concerned with regressions?

You mentioned yourself wanting to get these improvements in the 5.0 release. That release is due November 27th, which is two weeks from now. Which means any change we make between now and then has to be _solid_ and _stable_, and, well, not cause regressions. 

A regression could be:

- A theme that wants to be Gutenberg aware when 5.0 launches, and has perhaps optimized towards the current implementation of the Quote block, such as TwentyNineteen. (Though we'd probably be able to fix up _that_ specific theme beforehand, but the point is still valid for non twenty themes).
- Regarding the alignments, any existing post authored and saved before this change, when merged, might present an "invalid content" box, or show up as being authored in the Classic block, unless we create a deprecation handler. Creating a deprecation handler is possible, but it's not easy, and it's a relatively big thing to do which means it needs to be thoroughly reviewed and tested, which is hard given there are a lot of other open isssues in the [milestones](https://github.com/WordPress/gutenberg/milestones) we need to fix for the 5.0 release.
- If a left/right change is made, it could cause a regression for RTL languages where a different default might be wanted.
- If a change to margins is made, it could cause display issues on mobile, or in wide/full-wide configurations. For example, if you put a quote inside a full-wide columns block 

This is not to say any of those regressions can't be caught and addressed in this PR. It's just that if this PR has a goal of getting merged before the 5.0 release, then it will be held to a higher standard (and should probably be reduced in scope) than if, for example, it goes into the 5.1 release, at which point it's totally fair to expand the scope.

[the remark this answers is not in the document]
Comment (colorful-tones) [reply packages/block-library/src/quote/editor.scss]: @jasmussen thanks for the clarification. Makes absolute sense.

File: packages/block-library/src/quote/editor.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (sarahmonster) [reply packages/block-library/src/quote/editor.scss]: There isn't anything *complicated*, per se, about using pseudo-elements in CSS, but it presents a bit of cognitive dissonance when the three alignments are styled using a different approach. So for example, let's say in my theme I just want to change the colour of the border to be pink.. all I should need to do is apply `border-color: pink` to the `.is-large-quote` element in my theme, and I'm done. 

I'm not opposed to adding extra complexity if it adds value, but from a purely visual perspective, I'm not sure either solution is likely to be better for the majority of users.

File: packages/block-library/src/quote/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (sarahmonster) [reply packages/block-library/src/quote/index.js]: Thanks for digging into it and providing additional context! 

It looks like we're sometimes using a data-attribute for alignments, unless I'm misunderstanding it: 
https://github.com/WordPress/gutenberg/blob/085a6058eb6cb2c09470dfb1b5fab47b826c48b1/packages/block-library/src/table/editor.scss
https://github.com/WordPress/gutenberg/blob/3a210ab5c127c2a9b9791827515cae3a59452a20/packages/block-library/src/pullquote/editor.scss
(lots more too!)

Would it make sense to use that here as well?

File: packages/block-library/src/quote/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (talldan) [reply packages/block-library/src/quote/index.js]: I agree that there should be no class by default. It seems like the css way to do things in terms of this class being a 'modifier', and modifiers only being applied for a variation of a default. 

I also think the classes could be more verbose, e.g. 'is-right-aligned', 'is-center-aligned'.

**Review by talldan — CHANGES_REQUESTED** [review]
Hi @colorful-tones - this looks really good visually, thanks for working on it.

In my testing I spotted an issue. I opened up a post with some existing quotes that had text alignment. I ended up with some block validation issues and the existing content wasn't aligned as expected. Looks like a deprecation for the block is required. Once that's done it'd be good to test whether it resolves all those issues or there's more that needs to be done in terms of compatibility.

Here's what my post looked like originally:
screen shot 2018-12-04 at 2 53 39 pm

Here's how the same post looked when I switched to this branch:
screen shot 2018-12-04 at 2 57 10 pm

Newly added blocks worked well though. 👍

**Comment by colorful-tones** [comment]
I'm closing this out in order to re-focus the overall Pull Request for this.

Thanks for all the excellent feedback.
